### PR TITLE
SegmentationMetricsByPixels tests

### PR DESCRIPTION
**Pull request by StepanGoryachev**

### CODE REVIEW

File: batchflow/tests/segmentation_metrics_by_pixels_test.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+"""
+# pylint: disable=import-error, no-name-in-module
+from itertools import chain
+import numpy as np
+import pytest
+
+from batchflow.models.metrics import SegmentationMetricsByPixels
+
+# Dict {metric_alias : metric_names}
+METRICS_DICT = {'tpr' : ['true_positive_rate', 'sensitivity', 'recall'],
```
Comment (AlexeyKozhevin) [inline batchflow/tests/segmentation_metrics_by_pixels_test.py]: Let's add aliases for metrics into metrics classes. It will be easier to write something like 
`metrics.evaluate('fpr')` then `metrics.evaluate('false_positive_rate')`

File: batchflow/models/metrics/classify.py [hunk 2]
```diff
@@ -133,6 +132,20 @@ def __init__(self, targets, predictions, fmt='proba', num_classes=None, axis=Non
     def confusion_matrix(self):
         return self._confusion_matrix.sum(axis=0)
 
+    def infmean(self, arr):
+        """
+        We need to average array while skipping infs when there is at least one finite number along averaging axis.
```
Comment (SergeyTsimfer) [inline batchflow/models/metrics/classify.py]: Change doc to imperative style, e.g. `Compute the arithmetic mean along the axis, ignoring infs. `

File: batchflow/tests/segmentation_metrics_by_pixels_test.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+# Set predictions as 'labels'.
+LABELS = np.array([0, 1, 1, 0, 2, 0, 1, 1]).reshape(BATCH_SIZE, IMAGE_SIZE, IMAGE_SIZE)
+# Onehots are basically like probas, just with all 0 and a single 1.
+PROBA = np.eye(NUM_CLASSES)[LABELS]
+# Logit function gives ±infs on degenerate case of 0s and 1s, but it's okay for sigmoid function.
+LOGITS = np.log(PROBA / (1. - PROBA))
+# First param stands for predictions variable, second — for predictions type, third — for axis with class info.
+PREDICTIONS = [(LABELS, 'labels', None),
+               (PROBA, 'proba', 3),
+               (LOGITS, 'logits', 3)]
```
Comment (AlexeyKozhevin) [inline batchflow/tests/segmentation_metrics_by_pixels_test.py]: Let's also check different class axes.

Comment (AlexeyKozhevin) [reply batchflow/tests/segmentation_metrics_by_pixels_test.py]: And thresholds.

File: batchflow/models/metrics/classify.py [hunk 4]
```diff
@@ -297,7 +310,11 @@ def specificity(self, *args, **kwargs):
         return self.true_negative_rate(*args, **kwargs)
 
     def prevalence(self, *args, **kwargs):
-        return self._calc_agg(self.condition_positive, self.total_population, *args, **kwargs)
+        """
+        Parameter when_zero doesn't really matter in this case,
+        since total_population is never zero, when targets are not empty.
+        """
```
Comment (SergeyTsimfer) [inline batchflow/models/metrics/classify.py]: First line of doc should give a quick description of what the method does. Use `Notes` section for implementation/usage details, just like [guide](https://numpydoc.readthedocs.io/en/latest/format.html) prescribes

File: batchflow/tests/segmentation_metrics_by_pixels_test.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+                'fdr' : ['false_discovery_rate'],
+                'for' : ['false_omission_rate'],
+                'npv' : ['negative_predictive_value'],
+                'plr' : ['positive_likelihood_ratio'],
+                'nlr' : ['negative_likelihood_ratio'],
+                'dor' : ['diagnostics_odds_ratio'],
+                'fos' : ['f1_score', 'dice'],
+                'jac' : ['jaccard', 'iou']}
+# List all metrics' names from METRICS_DICT values.
+METRICS_LIST = list(chain.from_iterable(METRICS_DICT.values()))
```
Comment (SergeyTsimfer) [inline batchflow/tests/segmentation_metrics_by_pixels_test.py]: List comprehension can be cleaner

File: batchflow/tests/segmentation_metrics_by_pixels_test.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+               (PROBA, 'proba', 3),
+               (LOGITS, 'logits', 3)]
+
+class TestShape:
+    """
+    This class checks the shape of metrics' return value for all combinations of both aggregation types.
+
+    There is a following pattern in both tests:
+    0. Each function is preceded by data for it's parametrization.
+    1. Fixtures with preset parameters are applied.
```
Comment (SergeyTsimfer) [inline batchflow/tests/segmentation_metrics_by_pixels_test.py]: There are no fixtures used

File: batchflow/tests/segmentation_metrics_by_pixels_test.py [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+            Cross-batch aggregation type
+
+        multi_agg : string
+            Multiclass agregation type
+
+        exp_shape : None or tuple
+            Expected return value shape
+        """
+        metric = SegmentationMetricsByPixels(TARGETS, predictions, fmt, NUM_CLASSES, axis)
+        for metric_name in METRICS_LIST:
```
Comment (SergeyTsimfer) [inline batchflow/tests/segmentation_metrics_by_pixels_test.py]: If the test fails before assert (e.g. on metric evaluation) it would be quite hard to find on which metric exactly did it fail. You can make another parameter `metric_name` and get rid of loop inside the function.

File: batchflow/tests/segmentation_metrics_by_pixels_test.py [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+
+        assert res_shape == exp_shape
+
+class TestResult:
+    """
+    This class checks the contents of metrics' return value for all combinations of both aggregation types.
+
+    There is a following pattern in both tests:
+    0. Each function is preceded by data for it's parametrization.
+    1. Fixtures with preset parameters are applied.
```
Comment (SergeyTsimfer) [inline batchflow/tests/segmentation_metrics_by_pixels_test.py]: Same

File: batchflow/tests/segmentation_metrics_by_pixels_test.py [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+    2. Instance of SegmentationMetricsByPixels is being created.
+    3. Metric is being evaluated with given parameters.
+    4. It's result is being compared with expected one.
+    """
+
+    # First param stands for batch aggregation type, second — for multiclass one,
+    # third represents expected output contents for each type of metrics.
+    params = [(None, None, {'tpr' : np.array([1.00, 1.00, 0.00, 0.50, 1.00, 1.00]),
```
Comment (SergeyTsimfer) [inline batchflow/tests/segmentation_metrics_by_pixels_test.py]: Can you add some comments about how exactly did you get all these values?

File: batchflow/tests/segmentation_metrics_by_pixels_test.py (hunk 1 above)
Comment (StepanGoryachev) [reply batchflow/tests/segmentation_metrics_by_pixels_test.py]: `METRICS_ALIASES` added to `classify.py`

File: batchflow/tests/segmentation_metrics_by_pixels_test.py (hunk 3 above)
Comment (StepanGoryachev) [reply batchflow/tests/segmentation_metrics_by_pixels_test.py]: Axes are now also tested for 'channels_first' data format.

File: batchflow/models/metrics/classify.py (hunk 2 above)
Comment (StepanGoryachev) [reply batchflow/models/metrics/classify.py]: Fixed.
